### PR TITLE
[Merged by Bors] - feat: two topology lemmas

### PR DESCRIPTION
**Pull request by grunweg**
From sphere-eversion; I'm just upstreaming it.

Co-authored-by: @fpvandoorn

### CODE REVIEW

File: Mathlib/Topology/Compactness/Compact.lean [hunk 1]
```diff
@@ -935,6 +935,11 @@ theorem ClosedEmbedding.isCompact_preimage {f : X → Y} (hf : ClosedEmbedding f
   hf.toInducing.isCompact_preimage (hf.closed_range) hK
 #align closed_embedding.is_compact_preimage ClosedEmbedding.isCompact_preimage
 
+/-- The preimage of a compact set in the image of an open embedding is compact. -/
+lemma OpenEmbedding.isCompact_preimage {f : X → Y} (hf : OpenEmbedding f) {K : Set Y}
+    (hK: IsCompact K) (Kf : K ⊆ range f) : IsCompact (f ⁻¹' K) := by
+  rwa [hf.toInducing.isCompact_iff, image_preimage_eq_of_subset Kf]
```
Comment (alreadydone) [inline Mathlib/Topology/Compactness/Compact.lean:941]: Why not generalize this to `hf : Inducing f`?

File: Mathlib/Topology/PartialHomeomorph.lean [hunk 2]
```diff
@@ -1346,6 +1346,15 @@ noncomputable def toPartialHomeomorph [Nonempty α] : PartialHomeomorph α β :=
...
 #align open_embedding.to_local_homeomorph OpenEmbedding.toPartialHomeomorph
 
+/-- An embedding defines a homeomorphism to its range. -/
+noncomputable def _root_.Embedding.homeomorph : Homeomorph α (range f) := by
+  by_cases hX : Nonempty α
+  · exact (Homeomorph.Set.univ α).symm.trans <|
+      (h.toPartialHomeomorph f).homeomorphOfImageSubsetSource Subset.rfl image_univ
+  · rw [not_nonempty_iff] at hX
+    have : IsEmpty (range f) := by rw [isEmpty_coe_sort, range_eq_empty f]
+    exact Homeomorph.empty
```
Comment (alreadydone) [inline Mathlib/Topology/PartialHomeomorph.lean:1356]: This def isn't definitionally nice. I think the Equiv part should be given by [Equiv.ofInjective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/Set.html#Equiv.ofInjective), but you're free to split by cases etc. to prove continuity.

Comment (urkud) [reply Mathlib/Topology/PartialHomeomorph.lean:1356]: We have [Homeomorph.ofEmbedding](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Homeomorph.html#Homeomorph.ofEmbedding)

File: Mathlib/Topology/Compactness/Compact.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/Topology/Compactness/Compact.lean]: Generalised and upgraded. (I tried that, but got stuck for some reason.)
And yes, I saw the other version already. :-)

File: Mathlib/Topology/PartialHomeomorph.lean (hunk 2 above)
Comment (grunweg) [reply Mathlib/Topology/PartialHomeomorph.lean:1356]: Nice find. I've dropped that lemma.

**Comment by grunweg** [comment]
awaiting-review

File: Mathlib/Topology/Compactness/Compact.lean [hunk 3]
```diff
@@ -929,6 +929,17 @@ theorem Inducing.isCompact_preimage {f : X → Y} (hf : Inducing f) (hf' : IsClo
...
+/-- The preimage of a compact set in the image of an inducing map is compact. -/
+lemma Inducing.isCompact_preimage' {f : X → Y} (hf : Inducing f) {K : Set Y}
+    (hK: IsCompact K) (Kf : K ⊆ range f) : IsCompact (f ⁻¹' K) := by
+  rwa [hf.isCompact_iff, image_preimage_eq_of_subset Kf]
+
+lemma Inducing.isCompact_preimage_iff {f : X → Y} (hf : Inducing f) {K : Set Y}
+    (Kf : K ⊆ range f) : IsCompact (f ⁻¹' K) ↔ IsCompact K:= by
+  refine ⟨fun h ↦ ?_, fun hK ↦ isCompact_preimage' hf hK Kf⟩
+  rw [← image_preimage_eq_of_subset Kf]
+  exact h.image hf.continuous
```
Comment (urkud) [inline Mathlib/Topology/Compactness/Compact.lean:941]:
```suggestion
lemma Inducing.isCompact_preimage_iff {f : X → Y} (hf : Inducing f) {K : Set Y}
    (Kf : K ⊆ range f) : IsCompact (f ⁻¹' K) ↔ IsCompact K:= by
  rw [hf.isCompact_iff, image_preimage_eq_of_subset Kf]

/-- The preimage of a compact set in the image of an inducing map is compact. -/
lemma Inducing.isCompact_preimage' {f : X → Y} (hf : Inducing f) {K : Set Y}
    (hK: IsCompact K) (Kf : K ⊆ range f) : IsCompact (f ⁻¹' K) :=
  (hf.isCompact_preimage_iff Kf).2 hK

```

Comment (urkud) [reply Mathlib/Topology/Compactness/Compact.lean:941]: If you don't care about the order of arguments, then you may also use `alias` to generate the latter theorem (or drop it in favor of the first version).

**Comment by urkud** [comment]
bors d+

File: Mathlib/Topology/Compactness/Compact.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Topology/Compactness/Compact.lean]: Changed. I've kept the second lemma for now (that can happily be changed later).

**Comment by grunweg** [comment]
Thanks for the fast reviews!
bors r+

**Pull request closed without merging**